File: src/frontend/src/utils/wsLogger.js
/**
 * WebSocket Logger - Tracks payloads and latency for game waiting room
 * Usage: 
 *   wsLogger.ready('user-123', { type: 'player_ready', player_id: 123 })
 *   wsLogger.receive('game-invite-1-2-1234', { type: 'state', players: [...] })
 *   wsLogger.latency('ready_to_broadcast', startTime)
 */

const WS_LOG_ENABLED = true; // Set to false to disable all logging

class WebSocketLogger {
  constructor() {
    this.events = [];
    this.maxEvents = 100; // Keep last 100 events in memory
  }

  /**
   * Get current high-resolution timestamp in milliseconds
   */
  getTimestamp() {
    return performance.now();
  }

  /**
   * Format a log entry with styling for browser console
   */
  formatLog(level, category, message, data = null) {
    const timestamp = new Date().toISOString();
    const levelColor = {
      info: '#0066cc',
      error: '#cc0000',
      warn: '#ff9900',
      debug: '#666666',
    }[level] || '#000000';

    const prefix = `%c[${timestamp}] [${level.toUpperCase()}] [${category}]`;
    const args = [`${prefix}%c ${message}`, `color: ${levelColor}; font-weight: bold;`, 'color: #333'];

    if (data) {
      args.push(data);
    }

    return { prefix, args };
  }

  /**
   * Log ready button click with payload
   */
  ready(roomId, payload) {
    if (!WS_LOG_ENABLED) return;

    const entry = {
      type: 'ready_click',
      roomId,
      payload,
      timestamp: this.getTimestamp(),
      iso: new Date().toISOString(),
    };

    this.events.push(entry);
    if (this.events.length > this.maxEvents) this.events.shift();

    const { args } = this.formatLog('info', 'GameWaitingRoom', '🔘 Ready button clicked');
    console.log(...args, { roomId, payload, ts: entry.timestamp.toFixed(2) });
  }

  /**
   * Log WebSocket message sent
   */
  send(roomId, payload) {
    if (!WS_LOG_ENABLED) return;

    const entry = {
      type: 'send',
      roomId,
      payload,
      timestamp: this.getTimestamp(),
      iso: new Date().toISOString(),
    };

    this.events.push(entry);
    if (this.events.length > this.maxEvents) this.events.shift();

    const { args } = this.formatLog('info', 'WS Send', `↗️  Payload sent`, { roomId, payload });
    console.log(...args);
  }

  /**
   * Log WebSocket message received
   */
  receive(roomId, payload) {
    if (!WS_LOG_ENABLED) return;

    const entry = {
      type: 'receive',
      roomId,
      payload,
      timestamp: this.getTimestamp(),
      iso: new Date().toISOString(),
    };

    this.events.push(entry);
    if (this.events.length > this.maxEvents) this.events.shift();

    const { args } = this.formatLog('info', 'WS Receive', `↙️  Payload received`, {
      roomId,
      payload,
    });
    console.log(...args);
  }

  /**
   * Log connection state change
   */
  connection(roomId, state, metadata = {}) {
    if (!WS_LOG_ENABLED) return;

    const entry = {
      type: 'connection',
      roomId,
      state,
      metadata,
      timestamp: this.getTimestamp(),
      iso: new Date().toISOString(),
    };

    this.events.push(entry);
    if (this.events.length > this.maxEvents) this.events.shift();

    const icon = state === 'open' ? '🔗' : '🔌';
    const { args } = this.formatLog('debug', 'WS Connection', `${icon} ${state}`, {
      roomId,
      metadata,
    });
    console.log(...args);
  }

  /**
   * Log UI state update after receiving payload
   */
  uiUpdate(roomId, updateData) {
    if (!WS_LOG_ENABLED) return;

    const entry = {
      type: 'ui_update',
      roomId,
      updateData,
      timestamp: this.getTimestamp(),
      iso: new Date().toISOString(),
    };

    this.events.push(entry);
    if (this.events.length > this.maxEvents) this.events.shift();

    const { args } = this.formatLog('info', 'UI Update', `🎨 State updated`, {
      roomId,
      updateData,
    });
    console.log(...args);
  }

  /**
   * Calculate and log latency between two events
   * Usage: wsLogger.latency('ready_to_broadcast', readyClickTime)
   * Returns latency in milliseconds
   */
  latency(label, startTimestamp) {
    if (!WS_LOG_ENABLED) return null;

    const endTimestamp = this.getTimestamp();
    const latencyMs = (endTimestamp - startTimestamp).toFixed(2);

    const entry = {
      type: 'latency',
      label,
      latencyMs: parseFloat(latencyMs),
      startTimestamp,
      endTimestamp,
      iso: new Date().toISOString(),
    };

    this.events.push(entry);
    if (this.events.length > this.maxEvents) this.events.shift();

    const { args } = this.formatLog('info', 'Latency', `⏱️  ${label}`, {
      latencyMs: `${latencyMs}ms`,
    });
    console.log(...args);

    return parseFloat(latencyMs);
  }

  /**
   * Log complete flow (ready → send → receive → ui update)
   */
  flowStart(roomId, label) {
    if (!WS_LOG_ENABLED) return this.getTimestamp();

    const { args } = this.formatLog('debug', 'Flow', `▶️  Flow started: ${label}`, { roomId });
    console.log(...args);

    return this.getTimestamp();
  }

  flowEnd(roomId, label, startTime) {
    if (!WS_LOG_ENABLED) return null;

    const latency = this.latency(`${label} (complete flow)`, startTime);
    const { args } = this.formatLog('info', 'Flow', `✅ Flow completed: ${label}`, {
      totalLatencyMs: `${latency}ms`,
    });
    console.log(...args);

    return latency;
  }

  /**
   * Export all events for debugging
   */
  export() {
    return {
      exportedAt: new Date().toISOString(),
      eventCount: this.events.length,
      events: this.events,
    };
  }

  /**
   * Clear all logged events
   */
  clear() {
    this.events = [];
  }

  /**
   * Get summary statistics
   */
  summary() {
    const summary = {
      totalEvents: this.events.length,
      byType: {},
      timeSpan: null,
    };

    this.events.forEach((event) => {
      summary.byType[event.type] = (summary.byType[event.type] || 0) + 1;
    });

    if (this.events.length > 1) {
      const first = this.events[0].timestamp;
      const last = this.events[this.events.length - 1].timestamp;
      summary.timeSpan = `${(last - first).toFixed(2)}ms`;
    }

    return summary;
  }
}

// Create singleton instance
const wsLogger = new WebSocketLogger();

export default wsLogger;

File: src/backend/shared/logging/ws_logger.py
"""
WebSocket Logger for Backend - Tracks payloads and latency for game sessions
Usage:
    from shared.logging.ws_logger import ws_logger
    ws_logger.ready(game_id, player_id, payload)
    ws_logger.receive(game_id, player_id, payload)
    ws_logger.latency('ready_to_broadcast', start_time)
"""

import os
import logging
import time
import json
from typing import Any, Dict, Optional, List
from dataclasses import dataclass, asdict
from datetime import datetime

logger = logging.getLogger(__name__)

# Enable WS logging only if explicitly requested via environment variable (default: disabled for production)
# Automatically enabled when using `make seed` for development/debugging
# Can also be enabled manually: WS_LOG_DEBUG=true make re-back
WS_LOG_ENABLED = os.getenv('WS_LOG_DEBUG', 'false').lower() == 'true'


@dataclass
class WSLogEntry:
    """Represents a single WebSocket log entry"""
    type: str
    room_id: str
    timestamp: float
    iso_timestamp: str
    payload: Optional[Dict[str, Any]] = None
    player_id: Optional[int] = None
    metadata: Optional[Dict[str, Any]] = None
    latency_ms: Optional[float] = None


class WebSocketLogger:
    """Shared logging utility for WebSocket communication"""

    def __init__(self, max_events: int = 100):
        self.events: List[WSLogEntry] = []
        self.max_events = max_events

    def get_timestamp(self) -> float:
        """Get current timestamp in milliseconds"""
        return time.time() * 1000

    def get_iso_timestamp(self) -> str:
        """Get ISO 8601 formatted timestamp"""
        return datetime.utcnow().isoformat() + "Z"

    def _add_event(self, entry: WSLogEntry) -> None:
        """Add event to circular buffer"""
        self.events.append(entry)
        if len(self.events) > self.max_events:
            self.events.pop(0)

    def ready(
        self, game_id: str, player_id: int, payload: Dict[str, Any]
    ) -> None:
        """Log ready button click"""
        if not WS_LOG_ENABLED:
            return

        entry = WSLogEntry(
            type="ready_click",
            room_id=game_id,
            player_id=player_id,
            payload=payload,
            timestamp=self.get_timestamp(),
            iso_timestamp=self.get_iso_timestamp(),
        )

        self._add_event(entry)
        logger.info(
            f"[WS Ready] 🔘 Player {player_id} ready in game {game_id}: {json.dumps(payload)}"
        )

    def send(
        self, game_id: str, player_id: int, payload: Dict[str, Any]
    ) -> float:
        """Log message sent to client"""
        if not WS_LOG_ENABLED:
            return 0

        timestamp = self.get_timestamp()
        entry = WSLogEntry(
            type="send",
            room_id=game_id,
            player_id=player_id,
            payload=payload,
            timestamp=timestamp,
            iso_timestamp=self.get_iso_timestamp(),
        )

        self._add_event(entry)
        logger.debug(
            f"[WS Send] ↗️  Sent to player {player_id} in {game_id}: {json.dumps(payload)}"
        )

        return timestamp

    def broadcast(
        self, game_id: str, payload: Dict[str, Any], client_count: int = 0
    ) -> float:
        """Log message broadcasted to all clients"""
        if not WS_LOG_ENABLED:
            return 0

        timestamp = self.get_timestamp()
        entry = WSLogEntry(
            type="broadcast",
            room_id=game_id,
            payload=payload,
            timestamp=timestamp,
            iso_timestamp=self.get_iso_timestamp(),
            metadata={"client_count": client_count},
        )

        self._add_event(entry)
        logger.info(
            f"[WS Broadcast] 📡 Broadcasted to {client_count} clients in {game_id}: "
            f"{json.dumps(payload)}"
        )

        return timestamp

    def receive(
        self, game_id: str, player_id: int, payload: Dict[str, Any]
    ) -> float:
        """Log message received from client"""
        if not WS_LOG_ENABLED:
            return 0

        timestamp = self.get_timestamp()
        entry = WSLogEntry(
            type="receive",
            room_id=game_id,
            player_id=player_id,
            payload=payload,
            timestamp=timestamp,
            iso_timestamp=self.get_iso_timestamp(),
        )

        self._add_event(entry)
        logger.debug(
            f"[WS Receive] ↙️  Received from player {player_id} in {game_id}: "
            f"{json.dumps(payload)}"
        )

        return timestamp

    def connection(
        self, game_id: str, player_id: int, state: str, metadata: Optional[Dict] = None
    ) -> None:
        """Log connection state change (open/close/error)"""
        if not WS_LOG_ENABLED:
            return

        entry = WSLogEntry(
            type="connection",
            room_id=game_id,
            player_id=player_id,
            timestamp=self.get_timestamp(),
            iso_timestamp=self.get_iso_timestamp(),
            metadata=metadata or {},
        )

        self._add_event(entry)

        icon = "🔗" if state == "open" else "🔌"
        details = f" ({json.dumps(metadata)})" if metadata else ""
        logger.info(
            f"[WS Connection] {icon} Player {player_id} {state} in {game_id}{details}"
        )

    def session_state(self, game_id: str, state: Dict[str, Any]) -> None:
        """Log session state snapshot (players ready, score, etc)"""
        if not WS_LOG_ENABLED:
            return

        entry = WSLogEntry(
            type="session_state",
            room_id=game_id,
            payload=state,
            timestamp=self.get_timestamp(),
            iso_timestamp=self.get_iso_timestamp(),
        )

        self._add_event(entry)

        players_status = state.get("players", {})
        logger.debug(
            f"[WS Session] 🎮 Game {game_id} state: {json.dumps(state)}"
        )

    def latency(
        self, label: str, start_timestamp: float
    ) -> float:
        """Calculate and log latency between two points in time (in milliseconds)"""
        if not WS_LOG_ENABLED:
            return 0

        end_timestamp = self.get_timestamp()
        latency_ms = end_timestamp - start_timestamp

        entry = WSLogEntry(
            type="latency",
            room_id=label,
            timestamp=end_timestamp,
            iso_timestamp=self.get_iso_timestamp(),
            latency_ms=latency_ms,
        )

        self._add_event(entry)
        logger.info(
            f"[WS Latency] ⏱️  {label}: {latency_ms:.2f}ms"
        )

        return latency_ms

    def flow_start(self, game_id: str, label: str) -> float:
        """Log start of a flow (e.g., ready → broadcast → client update)"""
        if not WS_LOG_ENABLED:
            return 0

        timestamp = self.get_timestamp()
        logger.debug(f"[WS Flow] ▶️  Starting {label} in {game_id}")
        return timestamp

    def flow_end(self, game_id: str, label: str, start_timestamp: float) -> float:
        """Log completion of a flow with total latency"""
        if not WS_LOG_ENABLED:
            return 0

        latency_ms = self.latency(f"{label} (complete flow)", start_timestamp)
        logger.info(
            f"[WS Flow] ✅ Completed {label} in {game_id}: {latency_ms:.2f}ms total"
        )

        return latency_ms

    def error(self, game_id: str, player_id: Optional[int], message: str, exc: Optional[Exception] = None) -> None:
        """Log WebSocket errors"""
        if not WS_LOG_ENABLED:
            return

        entry = WSLogEntry(
            type="error",
            room_id=game_id,
            player_id=player_id,
            timestamp=self.get_timestamp(),
            iso_timestamp=self.get_iso_timestamp(),
            metadata={"message": message, "exception": str(exc) if exc else None},
        )

        self._add_event(entry)
        logger.error(f"[WS Error] ❌ Game {game_id} Player {player_id}: {message}", exc_info=exc)

    def export(self) -> Dict[str, Any]:
        """Export all logged events for debugging"""
        return {
            "exported_at": self.get_iso_timestamp(),
            "event_count": len(self.events),
            "events": [asdict(e) for e in self.events],
        }

    def clear(self) -> None:
        """Clear all logged events"""
        self.events.clear()
        logger.debug("[WS Logger] 🗑️  Cleared all events")

    def summary(self) -> Dict[str, Any]:
        """Get summary statistics of logged events"""
        summary = {
            "total_events": len(self.events),
            "by_type": {},
            "time_span_ms": None,
        }

        for event in self.events:
            summary["by_type"][event.type] = summary["by_type"].get(event.type, 0) + 1

        if len(self.events) > 1:
            first = self.events[0].timestamp
            last = self.events[-1].timestamp
            summary["time_span_ms"] = round(last - first, 2)

        return summary


# Create singleton instance
ws_logger = WebSocketLogger()
